src/hal/src/pso/descriptor.rs

```diff
@@ -87,26 +87,42 @@ pub struct DescriptorRangeDesc {
     pub count: usize,
 }
 
+/// An error allocating descriptor sets from a pool.
+#[derive(Fail, Clone, Copy, Debug, PartialEq)]
+pub enum AllocationError {
+    /// Memory allocation on the host side failed.
+    /// This could be caused by a lack of memory or pool fragmentation.
+    #[fail(display = "Host memory allocation failed.")]
+    OutOfHostMemory,
+    /// Memory allocation on the host side failed.
+    /// This could be caused by a lack of memory or pool fragmentation.
+    #[fail(display = "Device memory allocation failed.")]
+    OutOfDeviceMemory,
```

Comment:
The vulkan spec is unclear to me as to whether these errors mean something different to the fragmented pool error as I would assume it's possible to run out of memory without having a fragmented pool.